apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/PipelineApplicationState.java

```diff
@@ -0,0 +1,161 @@
+   *
+   * @param dataSource a non-pooled {@link DataSource} for the application's DB (which this will use
+   *     to create {@link #getPooledDataSource()})
+   * @param maxPoolSize the {@link DatabaseOptions#getMaxPoolSize()} value to use
+   * @param metrics the {@link MetricRegistry} to use
+   * @return a {@link HikariDataSource} for the BFD database
+   */
+  private static HikariDataSource createPooledDataSource(
+      DataSource dataSource, int maxPoolSize, MetricRegistry metrics) {
+    HikariDataSource pooledDataSource = new HikariDataSource();
+
+    pooledDataSource.setDataSource(dataSource);
+    pooledDataSource.setMaximumPoolSize(maxPoolSize);
+    pooledDataSource.setRegisterMbeans(true);
+    pooledDataSource.setMetricRegistry(metrics);
+
+    return pooledDataSource;
+  }
+
+  /**
+   * This is marked <code>public</code> only for use in tests; non-test code that needs a {@link
+   * EntityManagerFactory} should get it from the {@link PipelineApplicationState}.
+   *
+   * @param pooledDataSource the JDBC {@link DataSource} for the application's database
+   * @return a JPA {@link EntityManagerFactory} for the application's database
+   */
+  private static EntityManagerFactory createEntityManagerFactory(DataSource dataSource) {
+    /*
+     * The number of JDBC statements that will be queued/batched within a single transaction. Most
+     * recommendations suggest this should be 5-30. Paradoxically, setting it higher seems to
+     * actually slow things down. Presumably, it's delaying work that could be done earlier in a
+     * batch, and that starts to cost more than the extra network roundtrips.
+     */
+    int jdbcBatchSize = 10;
+
+    Map<String, Object> hibernateProperties = new HashMap<>();
+    hibernateProperties.put(org.hibernate.cfg.AvailableSettings.DATASOURCE, dataSource);
+    hibernateProperties.put(org.hibernate.cfg.AvailableSettings.HBM2DDL_AUTO, Action.VALIDATE);
+    hibernateProperties.put(
+        org.hibernate.cfg.AvailableSettings.STATEMENT_BATCH_SIZE, jdbcBatchSize);
+
+    EntityManagerFactory entityManagerFactory =
+        Persistence.createEntityManagerFactory("gov.cms.bfd", hibernateProperties);
+    return entityManagerFactory;
+  }
+
+  /** @return the {@link MetricRegistry} for the application */
+  public MetricRegistry getMetrics() {
+    return metrics;
+  }
+
+  /** @return the {@link HikariDataSource} that holds the application's database connection pool */
+  public HikariDataSource getPooledDataSource() {
+    return pooledDataSource;
+  }
+
+  /** @return the {@link EntityManagerFactory} for the application */
+  public EntityManagerFactory getEntityManagerFactory() {
+    return entityManagerFactory;
+  }
+
+  /** @see java.lang.AutoCloseable#close() */
+  @Override
+  public void close() throws Exception {
+    if (!pooledDataSource.isClosed()) {
+      pooledDataSource.close();
+    }
+    if (entityManagerFactory.isOpen()) {
+      entityManagerFactory.close();
+    }
```

Comment:
No idea if this matters, but, should the data source be closed after the entityManagerFactory just in case the latter needs it?  Maybe there is nothing to do in close() that needs it though.

Reply:
Great catch: fixed.